File: 0020.valid-parentheses/valid-parentheses.py
#括号匹配 大水题
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        result=[]
        slength=len(s)
        for c in s :
            if c =='(' or c=='{' or c=='[':
                result.append(c)
            else :
                if len(result)==0:
                    return False
                tail=result.pop()
                print (tail)
                if c==')' and tail=='(' or c=='}' and tail=='{' or c==']' and tail=='[' :
                    continue
                else :
                    return False
        if len(result)==0:
            return True
        else :
            return False

File: 0020.valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string ss) {
        stack<char> s;
        for(auto cc:ss){
            if(cc=='(') s.push(cc);
            else if(cc==')'){
                if(s.empty()) return false;
                if(s.top()!='(') return false;
                else s.pop();
            }
            else if(cc=='[') s.push(cc);
            else if(cc==']'){
                if(s.empty()) return false;
                if(s.top()!='[') return false;
                else s.pop();
            }
            else if(cc=='{') s.push(cc);
            else if(cc=='}'){
                if(s.empty()) return false;
                if(s.top()!='{') return false;
                else s.pop();
            }
            
        }
        if(!s.empty()) return false;
        return true;
    }
};